cmd/snap-repair/runner.go

```diff
@@ -285,11 +287,140 @@ func (run *Runner) initState() error {
 	}
 	// best-effort remove old
 	os.Remove(dirs.SnapRepairStateFile)
-	// TODO: init Device
+	if err := run.initDeviceInfo(); err != nil {
+		return err
+	}
 	run.stateModified = true
 	return run.SaveState()
 }
 
+func trustedBackstore(trusted []asserts.Assertion) asserts.Backstore {
+	trustedBS := asserts.NewMemoryBackstore()
+	for _, t := range trusted {
+		trustedBS.Put(t.Type(), t)
+	}
+	return trustedBS
+}
+
+func checkAuthorityID(a asserts.Assertion, trusted asserts.Backstore) error {
+	assertType := a.Type()
+	if assertType != asserts.AccountKeyType && assertType != asserts.AccountType {
+		return nil
+	}
+	// check that account and account-key assertions are signed by
+	// a trusted authority
+	acctID := a.AuthorityID()
+	_, err := trusted.Get(asserts.AccountType, []string{acctID}, asserts.AccountType.MaxSupportedFormat())
+	if err != nil && err != asserts.ErrNotFound {
+		return err
+	}
+	if err == asserts.ErrNotFound {
+		return fmt.Errorf("%v not signed by trusted authority: %s", a.Ref(), acctID)
+	}
+	return nil
+}
+
+func verifySignatures(a asserts.Assertion, workBS asserts.Backstore, trusted asserts.Backstore) error {
+	if err := checkAuthorityID(a, trusted); err != nil {
+		return err
+	}
+	acctKeyMaxSuppFormat := asserts.AccountKeyType.MaxSupportedFormat()
+
+	seen := make(map[string]bool)
+	bottom := false
+	for !bottom {
+		u := a.Ref().Unique()
+		if seen[u] {
+			return fmt.Errorf("circular assertions")
+		}
+		seen[u] = true
+		signKey := []string{a.SignKeyID()}
+		key, err := workBS.Get(asserts.AccountKeyType, signKey, acctKeyMaxSuppFormat)
+		if err != nil && err != asserts.ErrNotFound {
+			return err
+		}
+		if err == asserts.ErrNotFound {
+			key, err = trusted.Get(asserts.AccountKeyType, signKey, acctKeyMaxSuppFormat)
+			if err != nil && err != asserts.ErrNotFound {
+				return err
+			}
+			if err == asserts.ErrNotFound {
+				return fmt.Errorf("cannot find public key %q", signKey[0])
+			}
+			bottom = true
+		} else {
+			if err := checkAuthorityID(key, trusted); err != nil {
```

Comment:
I already thought about that and the did the change in the next PR in the sequence, I can move it here already though I think:

https://github.com/snapcore/snapd/pull/3616/commits/346c94eb5fac1a420a04430f804d8bd3499995bc